### PR TITLE
PrivxV2

### PR DESCRIPTION
**Pull request by ghostant-1017**

### CODE REVIEW

File: src/main.leo [hunk 1]
```diff
@@ -1,801 +1,437 @@
...
+        let next_id: u64 = Mapping::get_or_use(sequence, 0u64, 0u64);
+        let token_meta: TokenMeta = TokenMeta {
+            minter: caller,
+            token_id: next_id,
+            token_name: token_name,
+            token_symbol: token_symbol,
+            token_logo: token_logo,
+            circulating: 0u64,
+            total_supply: total_supply,
+            decimals: 6u8,
```
Comment (jinqiao0423) [inline src/main.leo:104]: hardcode?

File: src/main.leo [hunk 2]
```diff
@@ -1,801 +1,437 @@
...
+    finalize register(caller: address, token_name: u128, token_symbol: u64, token_logo: TokenLogo, total_supply: u64) {
+        let next_id: u64 = Mapping::get_or_use(sequence, 0u64, 0u64);
+        let token_meta: TokenMeta = TokenMeta {
+            minter: caller,
+            token_id: next_id,
+            token_name: token_name,
+            token_symbol: token_symbol,
+            token_logo: token_logo,
+            circulating: 0u64,
+            total_supply: total_supply,
```
Comment (jinqiao0423) [inline src/main.leo]: decimals is actually not used

File: src/main.leo [hunk 3]
```diff
@@ -1,801 +1,437 @@
...
-        let (real_quantity, token_fee): (u64, u64) = charge_fee(dealed_quantity);
-        let (real_amount, credits_fee): (u64, u64) = charge_fee(dealed_amount);
-        let buyer_return_amount: u64 = max_dealed_amount - dealed_amount;
-        // 1.Caculate token fee
-        let caller_token: u64 = Mapping::get_or_use(token2_deposited, caller_hash, 0u64);
-        Mapping::set(token2_deposited, caller_hash, caller_token + token_fee);
+        // 2.Ensure the token_id is matched
+        assert_eq(buy_order.token_id, sell_order.token_id);
+        // 3.Ensure the buy order's price > sell order's price
+        assert(buy_order.price > sell_order.price);
```
Comment (jinqiao0423) [inline src/main.leo:293]: >= ?

Comment (jinqiao0423) [inline src/main.leo]: buy_order.price >= sell_order.price ?

**Review by jinqiao0423 — APPROVED** [review]

**Review by jinqiao0423 — APPROVED** [review]

File: src/main.leo [hunk 4]
```diff
@@ -95,6 +105,10 @@ program privx_exchange.aleo {
         };
         Mapping::set(sequence, 0u64, next_id + 1u64);
         Mapping::set(registry, next_id, token_meta);
+        let token_setting: Setting = Setting {
+            claim_amount: 0u64,
+        };
+        Mapping::set(settings, 0u64, token_setting);
```
Comment (jinqiao0423) [inline src/main.leo]: Mapping::set(settings, next_id, token_setting);

File: src/main.leo [hunk 5]
```diff
@@ -74,7 +80,11 @@ program privx_exchange.aleo {
             total_supply: total_supply,
             decimals: 6u8,
         };
+        let token_setting: Setting = Setting {
+            claim_amount: 0u64,
```
Comment (jinqiao0423) [inline src/main.leo:84]: How about adding claim_amount as a parameter in transition register_stablecoin?

File: src/main.leo [hunk 6]
```diff
@@ -95,6 +105,10 @@ program privx_exchange.aleo {
         };
         Mapping::set(sequence, 0u64, next_id + 1u64);
         Mapping::set(registry, next_id, token_meta);
+        let token_setting: Setting = Setting {
+            claim_amount: 0u64,
```
Comment (jinqiao0423) [inline src/main.leo:109]: How about adding claim_amount as a parameter in transition register?

Comment (ghostant-1017) [reply src/main.leo:109]: Maybe add a extra function to change the settings is better

**Review by jinqiao0423 — APPROVED** [review]

**Review by jinqiao0423 — APPROVED** [review]

**Review by jinqiao0423 — APPROVED** [review]

**Pull request merged**